torch/_dynamo/graph_region_tracker.py

```diff
@@ -0,0 +1,225 @@
+import functools
+import math
+from collections import defaultdict, deque
+from typing import Any, Callable, Deque, Dict, List, Optional, Set, Tuple
+
+import torch.fx
+from torch.utils._pytree import tree_flatten
+
+
+Node = torch.fx.Node
+Region = List[Node]
+IdenticalNodes = List[Node]
+GlobalStateKey = Tuple[bool, bool, int, bool, bool, torch.dtype, bool, bool, bool]
+NodeKey = Tuple[GlobalStateKey, str, int]
+
+
+def _get_tensor_props(node: Node) -> Optional[Tuple[torch.Size, torch.dtype]]:
+    if isinstance(node, torch.fx.Node):
+        value = node.meta.get("example_value", None)
+        if isinstance(value, torch.Tensor):
+            return value.shape, value.dtype
+    return None
+
+
+@functools.lru_cache(128)
+def _hash_node_inputs(node: Node) -> int:
```

Comment:
The more I think about it, the more this problem looks like the AOTDisptacher warm caching problem. The shared problem is: given a Dynamo graph, some sample inputs, and some global state, how do we identify functionally identical graphs/sample inputs/global state that we see in the future?

Do we think there's anything worth sharing here?

Reply:
So I went through the AOT caching stuff, it looks like they just pickle the example value(s) directly, would this work with FakeTensor? It would be useful if we could reuse the input hashing strategy. 

The main issue I see with using the entire hashing strategy from the AOT warm cache though is that we are looking at adding a single node at a time (we don't have the region yet). In essence the current algorithm does what you describe, but instead of tagging nodes, we have a source/inputs key -> node mapping. To find the full regions we check nearby nodes that all have identical keys (ie sources) to add to all the regions in a region group.

Reply:
Actually I took another look, it looks like I can reuse the fake tensor reducer there to extract the tensor metadata, this will make this a lot easier.

Reply:
Sounds good